src/errors.ts

```diff
@@ -15,26 +15,43 @@ const isError = (err: any): err is ErrorDescription =>
   (err as ErrorDescription).error_description !== undefined &&
   typeof (err as ErrorDescription).error_description === "string";
 
-export const errorHandler = (req: Request, logger: Logger) => (
-  err?: unknown
-) => {
+const delay = (sleepMs: number): Promise<void> =>
+  new Promise((resolve) => setTimeout(resolve, sleepMs));
+
+type CustomError = Error & { nackDelayMs?: unknown };
+
+export const errorHandler = (
+  req: Request,
+  logger: Logger,
+  defaultDelayMs: number = 30000
+) => (err?: unknown): Promise<void> => {
   const requestInfo = {
     type: req.type,
     queue: req.queue,
-    properties: req.properties
+    properties: req.properties,
   };
 
   if (err instanceof Error) {
-    req.nack();
-    logger.error(err.stack ? err.stack : err.message, requestInfo);
+    const delayMs =
+      typeof (err as CustomError).nackDelayMs === "number"
+        ? ((err as CustomError).nackDelayMs as number)
+        : defaultDelayMs;
+    logger.error(err.stack ? err.stack : err.message, {
+      ...requestInfo,
+      delayMs,
+    });
+    return delay(delayMs).then(() => req.nack());
```

Comment:
Just to help my understanding of promises, how important is it that we return here?

Reply:
It decides when `callback` is called if someone does `mq.events(...).then(callback)`.

- If we don’t return anything in `errorHandler`, `callback` will be called _before_ `req.nack()`.
- If we return a promise in `errorHandler` (that resolves after the timeout), `callback` will be called _after_ `req.nack()`.

Also, if `req.nack()` were to throw an error:

- If we don’t return anything, that error would happen async and not be caught by anything.
- If we return a promise, that promise would be rejected (after the timeout and the error).

In practice, I don’t think either way matters much, but I think it’s cleaner to have the timeout part of the promise chain.